File: rpg_web/test/session-rules.test.js
import test from 'node:test';
import assert from 'node:assert/strict';

import { infusionDamage, infusionManaCost, magicArrowDamage, spectralInfusions, useInfusion, useMagicArrow, useSpell } from '../lib/classActions.js';
import { changeHitPoints, recordDeathSave } from '../lib/deathSaves.js';
import { emptyCharacter } from '../lib/rpgData.js';

test('três acertos contra a morte recuperam 1 de vida', () => {
  let character = { ...emptyCharacter(), maxHp: 20, currentHp: 0 };
  character = recordDeathSave(character, true);
  character = recordDeathSave(character, true);
  character = recordDeathSave(character, true);
  assert.equal(character.currentHp, 1);
  assert.equal(character.resources.deathSuccesses, 0);
});

test('cura limpa acertos, erros e estado morto', () => {
  const character = { ...emptyCharacter(), maxHp: 20, currentHp: 0, resources: { ...emptyCharacter().resources, deathSuccesses: 2, deathFailures: 3, dead: 1 } };
  const next = changeHitPoints(character, 4);
  assert.equal(next.currentHp, 4);
  assert.deepEqual([next.resources.deathSuccesses, next.resources.deathFailures, next.resources.dead], [0, 0, 0]);
});

test('infusão com 3 Cadência reduz PM e consome Foco', () => {
  const character = { ...emptyCharacter(), currentMana: 10, resources: { ...emptyCharacter().resources, focoCurrent: 8, cadenciaCurrent: 3 } };
  const result = useInfusion(character, spectralInfusions.find((item) => item.id === 'impact'));
  assert.equal(result.error, '');
  assert.equal(result.character.currentMana, 9);
  assert.equal(result.character.resources.focoCurrent, 7);
});

test('flechas mágicas calculam dano e penalidade de múltiplos ataques', () => {
  const impact = spectralInfusions.find((item) => item.id === 'impact');
  const spectral = spectralInfusions.find((item) => item.id === 'spectral');
  assert.deepEqual(infusionDamage(impact, 8, 3), { hit: 10, miss: 0 });
  assert.deepEqual(infusionDamage(spectral, 9, 0), { hit: 9, miss: 3 });
  assert.equal(infusionManaCost(impact, 3, 3), 2);
});

test('Flecha Mágica causa 2d4 e a infusão aprimora o mesmo disparo', () => {
  assert.deepEqual(magicArrowDamage(3, 4), { dice: [3, 4], total: 7 });
  const spell = { id: 'arrow', successfulUses: 0 };
  const character = { ...emptyCharacter(), spells: [spell] };
  const result = useMagicArrow(character, spell, { successful: true, dice: magicArrowDamage(2, 4) });
  assert.equal(result.damage.total, 6);
  assert.match(result.character.actionHistory[0].result, /2d4 = 2 \+ 4 = 6/);
  assert.equal(result.character.spells[0].successfulUses, 1);
});

test('magia registra terceiro sucesso e desconta recursos', () => {
  const spell = { id: 'spell', name: 'Seta astral', manaCost: 2, focusCost: 1, humanityCost: 3, successfulUses: 2 };
  const character = { ...emptyCharacter(), currentMana: 10, resources: { ...emptyCharacter().resources, focoCurrent: 5 }, spells: [spell] };
  const result = useSpell(character, spell, true);
  assert.equal(result.error, '');
  assert.equal(result.character.spells[0].successfulUses, 3);
  assert.equal(result.character.resources.humanity, 97);
});

test('magia divina registra o bônus de Fé dividido por dois', () => {
  const spell = { id: 'divine', name: 'Luz Julgadora', type: 'Divina', manaCost: 1, focusCost: 0, humanityCost: 0, successfulUses: 0 };
  const base = emptyCharacter();
  const character = { ...base, currentMana: 5, attributes: { ...base.attributes, faith: 7 }, resources: { ...base.resources, humanity: 40, divinity: 60 }, spells: [spell] };
  const result = useSpell(character, spell, true);
  assert.equal(result.damageBonus, 3);
  assert.match(result.character.actionHistory[0].result, /Fé \/ 2/);
});

File: rpg_sheet_app/test/session_rules_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:rpg_sheet_app/domain/calculators/proficiency_calculator.dart';
import 'package:rpg_sheet_app/domain/services/class_action_service.dart';
import 'package:rpg_sheet_app/domain/services/death_save_service.dart';
import 'package:rpg_sheet_app/models/character.dart';
import 'package:rpg_sheet_app/models/character_records.dart';

void main() {
  Character character({
    int hp = 20,
    int mana = 10,
    Map<String, int>? resources,
    List<String>? proficiencies,
  }) => Character(
    id: 'test',
    name: 'Teste',
    playerName: 'Jogador',
    raceId: 'race',
    classId: 'class',
    maxHp: 20,
    currentHp: hp,
    maxMana: 10,
    currentMana: mana,
    resources: resources,
    proficiencies: proficiencies,
  );

  test('proficiência adiciona novamente o valor bruto da perícia', () {
    const service = ProficiencyCalculator();
    expect(
      service.rawRollBonus(
        character(proficiencies: ['Furtividade']),
        'Furtividade',
        4,
      ),
      4,
    );
    expect(service.rawRollBonus(character(), 'Furtividade', 4), 0);
  });

  test('três acertos contra a morte recuperam 1 de vida', () {
    const service = DeathSaveService();
    var current = character(hp: 0);
    current = service.record(current, success: true);
    current = service.record(current, success: true);
    current = service.record(current, success: true);
    expect(current.currentHp, 1);
    expect(current.resources['deathSuccesses'], 0);
  });

  test('infusão desconta mana e foco com bônus de cadência', () {
    const service = ClassActionService();
    final result = service.useInfusion(
      character(
        resources: {'focoMax': 8, 'focoCurrent': 8, 'cadenciaCurrent': 3},
      ),
      ClassActionService.infusions[1],
    );
    expect(result.succeeded, isTrue);
    expect(result.character.currentMana, 9);
    expect(result.character.resources['focoCurrent'], 7);
    expect(result.character.actionHistory.single.result, contains('+1 efeito'));
  });

  test('flechas mágicas calculam dano e custo por múltiplos ataques', () {
    const service = ClassActionService();
    final impact = ClassActionService.infusions[1];
    final spectral = ClassActionService.infusions[4];
    expect(service.infusionDamage(impact, 8, 3).hit, 10);
    expect(service.infusionDamage(spectral, 9, 0).miss, 3);
    expect(service.infusionManaCost(impact, 3, 3), 2);
  });

  test('magia oficial preserva tópico e desconta mana e humanidade', () {
    const service = ClassActionService();
    final spell = CharacterSpell(
      id: 'spell_solar',
      catalogId: 'trello_solar',
      name: 'Sentença Solar',
      type: 'Divina',
      topic: 'Grau 1 · Utilidade e punição',
      manaCost: 1,
      humanityCost: 3,
      damage: '1d8 + Fé',
      createdAt: DateTime(2026),
    );
    final restored = CharacterSpell.fromJson(spell.toJson());
    final result = service.useSpell(
      character(resources: {'humanity': 100}).copyWith(spells: [restored]),
      restored,
      successful: true,
    );
    expect(restored.catalogId, 'trello_solar');
    expect(restored.topic, startsWith('Grau 1'));
    expect(result.character.currentMana, 9);
    expect(result.character.resources['humanity'], 97);
    expect(result.character.spells.single.successfulUses, 1);
  });
}
